### DISCUSSION TITLE
Whitespace rules

### DISCUSSION THREAD

**Discussion by JanEmde** [OP-TED/eForms-SDK#1072] [Ideas]
We have noticed that there are a few new rules regarding whitespace. The rule translation (at least in German and English) only mentions leading or trailing whitespaces. However, whitespaces between words are also recognised as invalid.  

Is that intentional? If so, the respective translations/descriptions of the rules should be adapted.

**Comment by JanEmde** [comment]
Any news?

**Reply by YvesJo** [reply]
Hi,
The check is indeed using the normalize-space function and therefore any leading, trailing or repeated space lead to an error.
The rules apply to codes, dates and identifiers, when one or more whitespaces, other than leading or trailing ones, occur in the field, the field would not satisfy additional requirements (e.g. Identifier format) and an additional rule would get triggered.
The change will need an update of the template used for the error messages. 
Thanks for the suggestion which will be discussed internally as soon as a major actor will be back.
KR,

**Reply by YvesJo** [reply]
If you would also have a text suggestion, please share it with us.
Thanks

**Reply by JanEmde** [reply]
My suggestion is:

German: ~'_Label_' (BT-XY) darf keine führenden, nachfolgenden oder doppelten Leerzeichen haben.~
'_Label_' (BT-XY) darf keine führenden, nachfolgenden oder aufeinanderfolgende Leerzeichen haben.
English:  '_Label_' (BT-XY) must not have leading, trailing or repeated spaces. or '_Label_' (BT-XY) must not have leading, trailing or repeated blanks.

**Reply by YvesJo** [reply]
Thanks

**Reply by rouschr** [reply]
I would suggest:
"'Label' (BT-XY) darf keine führende, nachfolgende oder aufeinanderfolgende Leerzeichen haben."
Doppelte means double / doubled which is technically not really translating the meaning of "repeated", **aufeinanderfolgende** ("on each other following") means consecutive.

**Reply by JanEmde** [reply]
You are right. "'Label' (BT-XY) darf keine führenden, nachfolgenden oder aufeinanderfolgende Leerzeichen haben." is better.

**Reply by rouschr** [reply]
And using the Latin "konsekutive" would work but that would be cheating ;)

**Comment by JanEmde** [comment]
@YvesJo We have noticed that there is no corresponding EFX-Rule for these Whitespace Schematron-Rules. This leads to problems in production and should be fixed as soon as possible.

**Reply by YvesJo** [reply]
Hi,
Could you please clarify what the last concern is? Is it about the rules labels, about the rules themselves or about the absence of EFX expression of the rules?

**Reply by JanEmde** [reply]
The absence of EFX expression of the rules.

**Reply by bertrand-lorentz** [reply]
Hello,
The rules on whitespace are indeed not expressed in EFX, because there is currently no function defined in EFX that is equivalent to the `normalize-space()` XPath function.
The Schematron for those rules is generated automatically for all fields that are of type "id", "id-ref", or "code" and do not correspond to XML attributes.
See this post that was published to gather feedback before this change was implemented: https://github.com/OP-TED/eForms-SDK/discussions/779

Could you explain a bit more about the problems you are facing related to this ?
Is it because your  users type in the values themselves, and the errors are only noticed during the submission of the XML notice ?
(For example, in eNotices2 the values for those fields are either created by the application, or selected in a list, so no extra whitespace can be added by the user.)

**Reply by JanEmde** [reply]
"Is it because your users type in the values themselves, and the errors are only noticed during the submission of the XML notice ?" This is exactly the problem.

BT-150-contract, for example, is specifically affected. What value is automatically assigned here by eNotices2?

**Reply by bertrand-lorentz** [reply]
I had missed BT-150-contract, it is indeed a text field in eNotices2, so the user can type in something that will cause the whitespace rule to fail.

We'll have to think above we can improve this, but it might be only possible with SDK 2.

**Reply by JanEmde** [reply]
At least BT-3201-Tender is also affected. 

These types of errors (absence of EFX expression of the rules) generate an enormous support effort and are therefore very annoying.

**Reply by JulianMa** [reply]
We just need an assertion, that there are no (leading and  trailing) whitespaces inside this particular Field, which could be done by emitting an additional assertion to (in this example) BT-150-Contract, which can be done in EFX 1 by the Like keyword (context might be wrong and regex not tested, just an example for leading and trailing):
```json
 "assert" : {
      "value" : true,
      "severity" : "ERROR",
      "constraints" : [ {
        "value" : "{ND-Root} ${BT-150-Contract like '^[^\s].+[^\s]'}",
        "severity" : "ERROR",
        "message" : "rule|text|BR-BT-XXXXX-XXXX",
      } ]
    },
```
and 
`rule|text|BR-BT-XXXXX-XXXX=BT-150-Contract must have no leading or trailing whitespaces`

**Reply by JanEmde** [reply]
> I had missed BT-150-contract, it is indeed a text field in eNotices2, so the user can type in something that will cause the whitespace rule to fail.
> 
> We'll have to think above we can improve this, but it might be only possible with SDK 2.

@bertrand-lorentz Would it be possible to remove the whitespace rules for the BT-150-contract and BT-3201-Tender fields? At least until there is an EFX solution for the problem?

**Comment by ferrakaop** [comment]
@GarrusMD we have looked at this issue and propose to remove the check for extra spaces within the field (so the rule would actually align with the error message). But I believe it's better to reject identifier fields that have leading or trailing whitespace. Even if these fields are entered manually by end users, it is very likely that any spaces are the start or the end are a mistake (often from copy and pasting), which we should pick up to be able to have better data quality.

Do you have any cases where something like a contract idenfitier should start or end with a space?

**Reply by JanEmde** [reply]
No, we don't have any problems with that. We trim out spaces at the beginning or end anyway. The problem only exists with repeated spaces within the strings.

**Reply by ferrakaop** [reply]
Thanks @GarrusMD for your feedback. We will adjust the rule in SDK 1.14 so that the repeated spaces within strings are no longer rejected - so the rule will match the existing error messages that only refers to leading or training spaces.
